File: v1/src/main/java/com/api/v1/borrower/services/DeleteBorrowerBySsnServiceImpl.java
package com.api.v1.borrower.services;

import com.api.v1.annotations.SSN;
import com.api.v1.borrower.domain.BorrowerRepository;
import com.api.v1.borrower.utils.BorrowerFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class DeleteBorrowerBySsnServiceImpl implements DeleteBorrowerBySsnService {

    @Autowired
    private BorrowerFinderUtil borrowerFinderUtil;

    @Autowired
    private BorrowerRepository repository;

    @Override
    public Mono<Void> deleteBySsn(@SSN String ssn) {
        return borrowerFinderUtil
                .find(ssn)
                .flatMap(borrower -> repository
                        .findAll()
                        .filter(e -> e.getSsn().equals(borrower.getSsn()))
                        .flatMap(repository::delete)
                        .then());
    }

}

File: koltlin/v1/src/main/kotlin/com/api/v1/borrower/services/DeleteBorrowerBySsnServiceImpl.kt
package com.api.v1.borrower.services

import com.api.v1.borrower.domain.Borrower
import com.api.v1.borrower.domain.BorrowerRepository
import com.api.v1.borrower.helpers.BorrowerFinderUtil
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Size
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class DeleteBorrowerBySsnServiceImpl: DeleteBorrowerBySsnService {

    @Autowired
    private lateinit var repository: BorrowerRepository

    @Autowired
    private lateinit var finder: BorrowerFinderUtil

    override fun deleteBySsn(@NotNull @Size(min=9, max=9) ssn: String): Mono<Void> {
        return finder.find(ssn).flatMap { b -> Mono.defer { repository.delete(b) } }
    }

}